lib/compiler/src/beam_ssa_alias.erl

```diff
@@ -0,0 +1,1000 @@
+             raw_raise ->
+                 SS1;
+             recv_marker_bind ->
+                 SS1;
+             recv_marker_clear ->
+                 SS1;
+             recv_marker_reserve ->
+                 SS1;
+             recv_next ->
+                 SS1;
+             remove_message ->
+                 SS1;
+             resume ->
+                 SS1;
+             update_tuple ->
+                 aa_construct_term(Dst, Args, SS1, AAS);
+             update_record ->
+                 [_Hint,_Size,Src|Updates] = Args,
+                 Values = [Src|aa_update_record_get_vars(Updates)],
+                 aa_construct_term(Dst, Values, SS1, AAS);
+             wait_timeout ->
+                 SS1;
+             _ ->
+                 exit({unknown_instruction, I})
+         end,
+    aa_is(Is, SS, [aa_update_annotation(I, SS1)|Acc],
+          AAS);
+aa_is([], SS, Acc, _AAS) ->
+    {reverse(Acc), SS}.
+
+aa_terminator(T=#b_br{anno=Anno0}, SS0) ->
+    Anno = aa_update_annotation(Anno0, SS0),
+    {T#b_br{anno=Anno}, SS0};
+aa_terminator(T=#b_ret{arg=Arg,anno=Anno0}, SS0) ->
+    Type = maps:get(result_type, Anno0, any),
+    Status0 = aa_get_status(Arg, SS0),
+    ?DP("Returned ~p:~p:~p~n", [Arg, Status0, Type]),
+    Type2Status0 = maps:get(returns, SS0, #{}),
+    Status = case Type2Status0 of
+                 #{ Type := OtherStatus } ->
+                     aa_meet(Status0, OtherStatus);
+                 #{ } ->
+                     Status0
+             end,
+    Type2Status = Type2Status0#{ Type => Status },
+    ?DP("new status map: ~p~n", [Type2Status]),
+    SS = SS0#{ returns => Type2Status},
+    {aa_update_annotation(T, SS), SS};
+aa_terminator(T=#b_switch{anno=Anno0}, SS0) ->
+    Anno = aa_update_annotation(Anno0, SS0),
+    {T#b_switch{anno=Anno}, SS0}.
+
+%% Add a new ssa variable to the alias state and set its status.
+aa_new_ssa_var(Var, Status, State) ->
+    false = maps:get(Var, State, false), % Assertion
+    State#{Var => {status, Status}}.
+
+aa_get_representative(Var, State) ->
+    %% TODO: Consider path compression
+    case State of
+        #{ Var := {status, _} } ->
+            Var;
+        #{ Var := Parent } ->
+            aa_get_representative(Parent, State);
+        #{} ->
+            Var
+    end.
+
+aa_get_status(V=#b_var{}, State) ->
+    Repr = aa_get_representative(V, State),
+    {status, S} = map_get(Repr, State),
+    S;
+aa_get_status(#b_literal{}, _State) ->
+    unique.
+
+%% As aa_get_status/2 but return false if nothing is known about the
+%% variable.
+aa_check_status(V=#b_var{}, State) ->
+    Repr = aa_get_representative(V, State),
+    case State of
+        #{ Repr := {status, S}} -> S;
+        #{} -> false
+    end.
+
+aa_set_status(V=#b_var{}, Status, State) ->
+    Repr = aa_get_representative(V, State),
+    State#{ Repr => {status, Status} };
+aa_set_status(#b_local{}, _Status, State) ->
+    State;
+aa_set_status(#b_remote{}, _Status, State) ->
+    State;
```

Comment:
Not covered by our tests. Can this happen? Is there a test case missing?

Reply:
I can't create an example which covers it, so away it goes! (fix in beb1977950b934fbd926bc223208c3ed2df3c895)